src/cython_cmake/cmake/UseCython.cmake

```
@@ -140,190 +72,120 @@
-    "${_options}"
-    "${_one_value}"
-    "${_multi_value}"
-    ${ARGN}
+function(Cython_compile_pyx)
+  cmake_parse_arguments(
+    PARSE_ARGV 0
+    CYTHON
+    ""
+    "OUTPUT;LANGUAGE;OUTPUT_VARIABLE"
+    "CYTHON_ARGS"
     )
-
-  # Configuration options.
-  set(CYTHON_ANNOTATE OFF
-      CACHE BOOL "Create an annotated .html file when compiling *.pyx.")
-
-  set(CYTHON_FLAGS "" CACHE STRING
-      "Extra flags to the cython compiler.")
-  mark_as_advanced(CYTHON_ANNOTATE CYTHON_FLAGS)
-
-  if(_args_C)
-    set(_target_language "C")
-  endif()
-  if(_args_CXX)
-    set(_target_language "CXX")
-  endif()
-
-  if(_args_PY2)
-    set(_language_level "2")
+  set(ALL_INPUT ${CYTHON_UNPARSED_ARGUMENTS})
+  list(LENGTH ALL_INPUT INPUT_LENGTH)
+  if(NOT INPUT_LENGTH EQUAL 1)
+    message(FATAL_ERROR "One and only one input file must be specified, got '${ALL_INPUT}'")
   endif()
-  if(_args_PY3)
-    set(_language_level "3")
-  endif()
-
-  list(GET _args_UNPARSED_ARGUMENTS 0 _arg0)
-
-  # if provided, use _arg0 as the input file path
-  if(_arg0)
-    set(_source_file ${_arg0})
-
-  # otherwise, must determine source file from name, or vice versa
-  else()
-    get_filename_component(_name_ext "${_name}" EXT)
+  list(GET ALL_INPUT 0 INPUT)
 
-    # if extension provided, _name is the source file
-    if(_name_ext)
-      set(_source_file ${_name})
-      get_filename_component(_name "${_source_file}" NAME_WE)
-
-    # otherwise, assume the source file is ${_name}.pyx
+  # Set target language
+  if(NOT CYTHON_LANGUAGE)
+    get_property(_languages GLOBAL PROPERTY ENABLED_LANGUAGES)
+
+    if("C" IN_LIST _langauges AND "CXX" IN_LIST _languages)
+      # Try to compute language. Returns falsy if not found.
+      _cython_compute_language(CYTHON_LANGUAGE ${INPUT})
+      message(STATUS "${CYTHON_LANGUAGE}")
+    elseif("C" IN_LIST _languages)
+      # If only C is enabled globally, assume C
+      set(CYTHON_LANGUAGE C)
+    elseif("CXX" IN_LIST _languages)
+      # Likewise for CXX
+      set(CYTHON_LANGUAGE "CXX")
     else()
-      set(_source_file ${_name}.pyx)
+      message(FATAL_ERROR "LANGUAGE keyword required if neither C nor CXX enabled globally")
     endif()
   endif()
 
-  # Set additional flags.
-  set(_cython_args)
-  if(CYTHON_ANNOTATE)
-    list(APPEND _cython_args "--annotate")
-  endif()
-  if(CMAKE_BUILD_TYPE STREQUAL "Debug" OR
-      CMAKE_BUILD_TYPE STREQUAL "RelWithDebInfo")
-    list(APPEND _cython_args
-      "--gdb"
-      "--line-directives"
-    )
-  endif()
-  string(STRIP "${CYTHON_FLAGS}" _stripped_cython_flags)
-  if(_stripped_cython_flags)
-    string(REGEX REPLACE " " ";" CYTHON_FLAGS_LIST "${_stripped_cython_flags}")
-    list(APPEND _cython_args ${CYTHON_FLAGS_LIST})
+  if(NOT CYTHON_LANGUAGE)
+    message(FATAL_ERROR "LANGUAGE keyword or `# distutils: language=...` required if C and CXX are enabled globally")
+  elseif(CYTHON_LANGUAGE STREQUAL C)
+    set(language_arg "")
+    set(langauge_ext ".c")
```

Needs fixing elsewhere too, find-and-replace in an editor will be easier than in the GUI though.
```suggestion
    set(language_ext ".c")
```